teps/0117-tekton-results-logs.md

```diff
@@ -0,0 +1,364 @@
+storing the logs in its configured log backend. After the logs are stored, the
+associated `TaskRunLog` record should be updated to record where the logs can
+be retrieved.
+
+The following storage categories may also be returned:
+
+- `None` - this indicates that the APIServer is not configured to store
+  forwarded logs.
+- `External` - this indicates that an external program or service is
+  responsible for forwarding and storing the logs. The APIServer is configured
+  to retrieve the logs from an appropriate service or storage location.
+  For example, clusters can be configured to forward Tekton `TaskRun` logs with
+  [fluentd](https://www.fluentd.org/) to
+  [ElasticSearch](https://github.com/elastic/elasticsearch). In this scenario,
+  log forwarding is delegated to fluentd, and the APIServer is configured to
+  retrieve logs from ElasticSearch.
+
+The Results APIServer will be extended with the following gRPC endpoints:
+
+- `GetLog` - this streams the logs associated with `TaskRunLog` record over
+  gRPC.
+- `ListLogs` - this lists all available logs for a given Result.
+- `PutLog` - this allows the watcher to stream logs associated with a
+  `TaskRunLog` record to the APIServer's storage backend.
+- `DeleteLog` - this deletes the log and its associated record.
+
+Like other Results APIs, the logs will be protected by subject access reviews
+and Kubernetes RBAC. A new resource type - "logs" - will be used to control
+access, with the following supported verbs:
+
+- `get` (for `GetLog`)
+- `list` (for `ListLogs`)
+- `update` (for `PutLog`)
+- `delete` (for `DeleteLog`)
+
+### Notes and Caveats
+
+- As a simplifying assumption, the the APIServer will only support a single log
+  backend type.
+- `TaskRunLog` metadata information is stored as a Record - this means that can
+  be retrieved using the Record APIs (`GetRecord`, `ListRecord`, etc.)
+- The watcher will support a configuration option to forward logs - by default
+  it will not forward logs.
+- The APIServer will have separate configuration options that configure the log
+  storage backend.
+
+
+## Design Details
+
+### TaskRunLog Object
+
+The `TaskRunLog` object has the look and feel of a Tekton object, which
+facilitates querying. This object exists as a Tekton Result record - it does
+not exist on the Kubernetes cluster itself.
+
+```yaml
+kind: TaskRunLog
+apiVersion: results.tekton.dev/v1alpha2
+# Standard k8s metadata - this should be copied from the related TaskRun
+metadata: 
+  name: <TaskRun Name> # Could match the related TaskRun
+  namespace: <TaskRun Namespace> # Could match the related TaskRun
+spec:
+  recordName: <Results record name>
+  taskRunRef:
+    name: <TaskRun Name>
+    namespace: <TaskRun Name>
+status: # All status fields are populated by the APIServer
+  size: 1024 # size of the log file, in bytes
+  storage:
+    category: None | Results | External
+    type: None | File | S3 | GCS | <additional backends>...
```

Comment:
For the external category, how is the storage location populated. Is that left up to the external provider to implement? Do we also need to configure some more information about the specific kind of external log forwarder?

Reply:
My thoughts here:

- The APIServer needs some level of configuration to know where to fetch the logs. Example - for Elasticsearch, the URL, index, and perhaps other info will need to be provided.
- The storage "provider" would need to populate the specific information needed to fetch the logs for a given TaskRun. It would be up to each provider to determine how this is accomplished.

Reply:
Yeah the second point seems like some kind of plugin (could be in-tree) that each external storage provider would need to provide